tests/unit/test_languageHandler.py

```diff
@@ -31,3 +98,119 @@ def test_nonStandardLocale(self):
 	def test_invalidLocale(self):
 		lcid = languageHandler.localeNameToWindowsLCID("zzzz")
 		self.assertEqual(lcid, LCID_NONE)
+
+
+class TestSetLocale(unittest.TestCase):
+	"""
+	tests setting the python locale for all possible locales set by NVDA or the System
+	"""
+
+	SUPPORTED_LOCALES = [("en", "en_US"), ("fa-IR", "fa_IR"), ("an-ES", "an_ES")]
+
+	@classmethod
+	def tearDownClass(cls):
+		languageHandler.setLanguage(languageHandler.curLang)
+
+	def setUp(self):
+		languageHandler.setLanguage(languageHandler.curLang)
+
+	def testSupportedLocales(self):
+		for localeName in self.SUPPORTED_LOCALES:
+			languageHandler.setLocale(localeName[0])
+			self.assertEqual(locale.getlocale()[0], localeName[1])
+
+	def testUnsupportedLocales(self):
+		original_locale = locale.getlocale()
+		for localeName in UNSUPPORTED_PYTHON_LOCALES:
+			with self.subTest():
+				languageHandler.setLocale(localeName)
+				self.assertEqual(locale.getlocale(), original_locale)
+
+	def testAllSupportedLangs(self):
+		for localeName in TRANSLATABLE_LANGS - UNSUPPORTED_PYTHON_LOCALES:
+			with self.subTest():
+				languageHandler.setLocale(localeName)
+				current_locale = locale.getlocale()
+
+				if localeName == "uk":
+					self.assertEqual(current_locale[0], "English_United Kingdom")
+				else:
+					pythonLang = current_locale[0].split("_")[0]
+					langOnly = localeName.split("_")[0]
+					self.assertEqual(
+						langOnly,
+						pythonLang,
+						f"full values: {localeName} {current_locale[0]}",
+					)
+
+	def testAllWindowsLangs(self):
+		prev_locale = locale.getlocale()
+		for localeName in WINDOWS_LANGS:
+			with self.subTest():
+				languageHandler.setLocale(localeName)
+				current_locale = locale.getlocale()
+				if localeName == languageHandler.curLang:
+					self.assertEqual(current_locale, prev_locale)
+				elif localeName in UNSUPPORTED_PYTHON_LOCALES:
+					self.assertEqual(current_locale, prev_locale)
+				else:
+					self.assertNotEqual(current_locale, prev_locale, localeName)
+
+
+class TestSetLanguage(unittest.TestCase):
+	"""
+	tests setting the NVDA language for all possible locales set by NVDA
+	"""
+	UNSUPPORTED_WIN_LANGUAGES = ["an", "kmr"]
+
+	def tearDown(self):
+		languageHandler.setLanguage(self._prevLang)
+
+	def __init__(self, *args, **kwargs):
+		self._prevLang = languageHandler.getLanguage()
+
+		ctypes.windll.kernel32.SetThreadLocale(0)
+		defaultThreadLocale = ctypes.windll.kernel32.GetThreadLocale()
+		self._defaultThreadLocaleName = languageHandler.windowsLCIDToLocaleName(
+			defaultThreadLocale
+		)
+
+		locale.setlocale(locale.LC_ALL, "")
+		self._defaultPythonLocale = locale.getlocale()
+
+		languageHandler.setLanguage(self._prevLang)
+		super().__init__(*args, **kwargs)
+
+	def testTranslatableLanguages(self):

```

Comment:
It might be worth considering the naming of these tests, from the name alone it's hard to differentiate them. A common way to name unit tests is `UnitOfWork_StateUnderTest_ExpectedBehavior`  see https://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html

We don't have a strong convention for this in NVDA tests, and this format is not always easy to follow, especially for "God objects" or functions that have very complex logic. These are however prime targets for refactoring.

Reply:
I've renamed these. I've included `UnitOfWork` but think it might be redundant here given the class names so let me know your thoughts

Reply:
Yes, I suppose it. It's not really a big deal. If you want to remove that, then I'd try to make the class names more clear EG
``` python
class test_languageHandler_setlocale
""" tests for the setlocale function in the languageHandler module."""
...

class test_languageHandler_setLanguage
""" tests for the setLanguage function in the languageHandler module."'"
...

```